slack-lint-checks/src/main/java/slack/lint/util/SlackJavaEvaluator.kt

```diff
@@ -0,0 +1,229 @@
+  // endregion
+
+  override fun hasModifier(owner: PsiModifierListOwner?, keyword: KtModifierKeywordToken): Boolean {
+    val superValue = super.hasModifier(owner, keyword)
+    // If it's not a compiled element or not a PsiClass, trust the super value and move on
+    if (!checkMetadata || owner !is PsiCompiledElement || owner !is PsiClass) {
+      return superValue
+    }
+
+    // We're working with an externally compiled element and it's a PsiClass, so we can do more
+    // thorough checks here.
+    KOTLIN_CLASS_MODIFIER_TOKENS[keyword]?.let { (flag, applicableClassKinds) ->
+      owner.findContaining(UClass::class.java)?.let { cls ->
+        // Only parse if the target class kind is applicable to the token we're checking. For
+        // example - when checking `data` tokens, they're not applicable to interfaces or enums.
+        if (cls.classKind in applicableClassKinds) {
+          cls.getOrParseMetadata()?.let { kmClass ->
+            return flag(kmClass.flags)
+          }
+        }
+      }
+    }
+
+    return superValue
+  }
+
+  private fun UAnnotated.getOrParseMetadata(): KmClass? {
+    val cls =
+      when (this) {
+        is UClass -> this
+        else -> return null // Only classes are supported right now
+      }
+    return cachedClasses
+      // Don't use getOrPut. Kotlin's extension may still invoke the body and we don't want that
+      .computeIfAbsent(qualifiedName!!) { key ->
+        val annotation =
+          cls.findAnnotation("kotlin.Metadata") ?: return@computeIfAbsent Optional.empty()
+        val (durationMillis, metadata) =
+          measureTimeMillisWithResult { annotation.parseMetadata(key) }
+        slackLintLog("Took ${durationMillis}ms to parse metadata for $key.")
+        Optional.ofNullable(metadata)
+      }
+      .getOrNull()
+  }
+
+  private fun UAnnotation.parseMetadata(classNameHint: String): KmClass? {
+    val metadataAnnotation =
+      MetadataWithNullableArgs(
+        kind = findAttributeValue("k")?.parseIntMember(),
+        metadataVersion = findAttributeValue("mv")?.parseIntArray(),
+        data1 = findAttributeValue("d1")?.parseStringArray(),
+        data2 = findAttributeValue("d2")?.parseStringArray(),
+        extraString = findAttributeValue("xs")?.parseStringMember(),
+        packageName = findAttributeValue("pn")?.parseStringMember(),
+        extraInt = findAttributeValue("xi")?.parseIntMember(),
+      )
+    val metadata = KotlinClassMetadata.read(metadataAnnotation)
+    return when (metadata) {
+      is KotlinClassMetadata.Class -> metadata.toKmClass()
+      is KotlinClassMetadata.FileFacade -> null
+      is KotlinClassMetadata.MultiFileClassFacade -> null
+      is KotlinClassMetadata.MultiFileClassPart -> null
+      is KotlinClassMetadata.SyntheticClass -> null
+      is KotlinClassMetadata.Unknown -> null
+      null -> {
+        slackLintLog("Could not load metadata for $classNameHint from file $file")
+        null
+      }
```

Comment:
There's a lot going on in this method. Consider extracting this `when` to a helper

Reply:
Cleaned it up in [26cdd85](https://github.com/slackhq/slack-lints/pull/75/commits/26cdd8535012d402b25fbca0f7120d09a4740010)